tortoise/README.md

```diff
@@ -0,0 +1,167 @@
+
+- against
+  
+    Against is added only if the base ballot supports the block and local opinion is against that block. 
+
+- abstain (undecided) 
+    
+    Undecided is a valid vote only until hare is terminated for the layer (it may take more than one layer for hare to terminate, this is expressed as zdist).
+
+If ballot doesn't specify explicit vote for a block then it votes agains the block. This prevents malicious smeshers from retroactively reorganizing mesh by intentionally creating late ballots.
+
+There are 4 distinct reasons to cast a vote:
+- hare output
+    
+    tortoise uses output from the hare consensus for voting on blocks within hdist.
+
+- validity/decisions
+
+    for layer before hdist the vote is casted according to the decision. 
+
+- local threshold
+
+    in case when decision wasn't reached but margin from counted weight crosses local threshold tortoise will cast a vote according to the local threshold.
+
+- weak coin
+
+    otherwise tortoise votes according to the coin recorded in the last layer. if sufficient number of honest votes were casted this way then tortoise will either reach decision or atleast vote according to the local threshold in future layers. if honest nodes disagreed on the coin value then the voting will continue until they agree.
+
+## Full tortoise
+
+Full tortoise counts votes from every ballot on every block before the ballot. Once counted weight cross global threshold tortoise can make a decision on a block. Layer is finalized only when all blocks have a decision.
+
+| ballot/block | weight | 0x11 | 0x22 | 0x33 | 0x44 |
+| ------------ | ------ | ---- | ---- | ---- | ---- |
+| 0xaa         | 10     | 1    | -1   | -1   | 1    |
+| 0xbb         | 20     | 1    | 1    | -1   | 1    |
+| total        | -      | 30   | 10   | -30  | 30   |
+| threshold    | -      | 20   | 20   | 20   | 20   |
+| decision     | -      | 1    | 0    | -1   | 1    |
+
+In the above example there are two ballots (0xaa and 0xbb). They vote for 4 blocks. Blocks 0x11 and 0x44 are valid and can be applied to the ledger, block 0x33 is invalid, and block 0x22 is still undecided.
+
+#### Votes from ballots with wrong beacon
+
+Because a smesher can select arbitrary beacon, there is an attack that will allow malicious smeshers to concentrate weight in a specific layer. If tortoise counts malicious weight immediatly it will make erroneous decisions (decide that a block is valid when in fact it is not). In order to prevent such attack tortoise will delay counting ballots with a wrong beacon until enough weight from honest smeshers was counted. 
+
+Note that we can't discard ballots with wrong beacon completely as recovery from partition requires counting ballots from both sides. And they will likely have different beacons if it was a long partition, or occured at the time when tortoise beacon was computed.
+
+#### Scaling issues
+
+Full tortoise complexity grows with the number of layers that are not finalized. 
```

Comment:
Why/how? Computational, memory, messages, or some other form of complexity?

Reply:
mostly computational, for every new ballot we have to update all blocks in all undecided layers